src/inmanta/protocol/ipc_light.py

```diff
@@ -246,3 +255,55 @@ def connection_lost(self, exc: Exception | None) -> None:
         super().connection_lost(exc)
         for fin in self.finalizers:
             asyncio.get_running_loop().create_task(fin())
+
+
+class LogReceiver(IPCFrameProtocol[ServerContext]):
+    """
+    IPC feature to receive log message
+    """
+
+    def frame_received(self, frame: IPCRequestFrame[ServerContext, ReturnType] | IPCReplyFrame) -> None:
+        if isinstance(frame, IPCLogRecord):
+            # calling log here is safe because if there are no arguments, formatter is never called
+            logging.getLogger(frame.name).log(frame.levelno, frame.msg)
+        else:
+            super().frame_received(frame)
+
+
+class LogShipper(logging.Handler):
+    """
+    Log sender associated with the log receiver
+
+    This sender is threadsafe
+    """
+
+    def __init__(self, protocol: IPCFrameProtocol[object], eventloop: asyncio.BaseEventLoop) -> None:
+        self.protocol = protocol
+        self.eventloop = eventloop
+        self.logger_name = "inmanta.ipc.logs"
+        self.logger = logging.getLogger(self.logger_name)
+        super().__init__()
+
+    def _send_frame(self, record: IPCLogRecord) -> None:
+        try:
+            self.protocol.send_frame(record)
+        except Exception:
+            # Stop exception here
+            # Log in own logger to prevent loops
+            self.logger.info("Could not send log line", exc_info=True)
+            return
+
+    def emit(self, record: logging.LogRecord) -> None:
+        if record.name == self.logger_name:
+            # avoid loops
+            return
```

Comment:
Should you elaborate on this. The log only flow in one direction. How would it be possible to get a loop?

Reply:
When the logger fails to send a log message, it creates a log message, ...